python/tvm/relay/backend/contrib/cmsisnn/codegen.py

```diff
@@ -0,0 +1,140 @@
+                return False
+            softmax_call = call.args[0]
+            if not self.call_contains_op(softmax_call, "nn.softmax"):
+                return False
+            dequantize_call = softmax_call.args[0]
+            if not self.call_contains_op(dequantize_call, "qnn.dequantize"):
+                return False
+            if not call.attrs.out_dtype == "int8":
+                return False
+            self.scale = dequantize_call.args[1].data.numpy().item(0)
+            return True
+
+        def emit_softmax_tir(self, call):
+            """Generates TIR extern_call for softmax"""
+            shape = call.checked_type.shape  # NHWC
+            dtype = call.checked_type.dtype
+            ir_builder = tvm.tir.ir_builder.create()
+            in_buf = tvm.tir.decl_buffer(shape=shape, dtype=dtype)
+            out_buf = tvm.tir.decl_buffer(shape=shape, dtype=dtype)
+            num_rows = shape[0] * shape[1] * shape[2]
+            row_size = shape[3]
+            ir_builder.emit(
+                tvm.tir.call_extern(
+                    dtype,
+                    "arm_softmax_s8",
+                    in_buf.data,
+                    num_rows,
+                    row_size,
+                    self.scale,
+                    out_buf.data,
+                )
+            )
+            prim_func = tvm.tir.PrimFunc([in_buf, out_buf], ir_builder.get())
+            prim_func = prim_func.with_attr("global_symbol", self.name)
+            prim_func = prim_func.with_attr("tir.noalias", True)
+            self.tir_mod = tvm.IRModule({self.name: prim_func})
+
+        def visit_call(self, call):
+            """Iterates over the relay operators within relay external function"""
+            super().visit_call(call)
+            if self.is_quantized_softmax(call):
+                self.emit_softmax_tir(call)
+
+        def generate_tir(self, func):
+            self.visit(func)
+            return self.tir_mod
+
+    tir_mod = GenerateTIR(name).generate_tir(func)
+    return tir_mod
+
+
+def relay_to_tir(name, func):
+    """Lower a Relay function to TIR for the CMSIS-NN target.
+
+    The Relay function should only contain operations supported
+    by the CMSIS-NN target. This is enforced by the graph partitioner
+    for CMSIS-NN.
+
+    Parameters
+    ----------
+    name: str
+        Name of the external relay function
+    func : tvm.relay.Function
+        The Relay function to lower.
+
+    Returns
+    -------
+    mod : tvm.IRModule
+        The lowered TIR module.
+
+    """
+    tir_mod = generate_tir(name, func)
+    return tir_mod
+
+
+@tvm._ffi.register_func("relay.ext.cmsisnn")
+def cmsisnn_compiler(relay_func):
+    """It compiles Relay's external function into equivalent TIR
+    and subsequently converts that into 'c' code. During the 'c'
+    code generation, it embeds CMSIS-NN APIs for the corresponding
+    operators.
+    """
+    assert isinstance(relay_func, tvm.ir.function.BaseFunc)
```

Comment:
What's the purpose of this assertion?

Reply:
Result of copy-n-paste. But my guess is that non AOT flows might be supplying other types too.

Reply:
I'm not convinced anything other than a function will come here, but if you're concerned can we have a test case that this assertion is raised?

Reply:
Ok, I will remove it then.